flax/linen/partitioning.py

```diff
@@ -261,14 +269,58 @@ def with_sharding_constraint(
       is_leaf=_is_logical_spec)
 
 
+# ------------------------------------------------------------------------------
+# Logical Partitioning Axis Metadata
+# ------------------------------------------------------------------------------
+
+
+class LogicallyPartitioned(meta.Partitioned):
```

Comment:
I like the Logical partitioning as an abstraction opt-in on top!

Maybe put it in a different module though? I think people will get confused about the old and new api being in a module together and having no interop. Also keeping this module as the old API will make it easier to deprecate it.

Reply:
Yeah, I can do that - the main reason I stuck it here is that I don't want to replicate the code handling the logical->mesh context, so I might move that into the new location and re-import it for the old file.